### PR TITLE
refactor: convert to typescript and upgrade Node.js versions

### PR DESCRIPTION
**Pull request by MarshallOfSound**
Converts to typescript so that forge can just import the options type instead of duplicating it.

This is a breaking change per the commit message, the export is now a named export `createDMG` and the `debug` option has been removed because it makes no sense.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+   */
+  name: string;
+  /**
+   * The title of the produced DMG, which will be shown when mounted.
+   *
+   * Defaults to the application name.
+   */
+  title?: string;
+  /**
+   * Path to the background for the DMG window. Background image should be of size 658x498.
```
Comment (erickzhao) [inline src/index.ts:24]:
```suggestion
   * Path to the background image for the DMG window. Image should be of size 658x498.
   * If you need to want to add a second Retina-compatible size, add a separate `@2x` image.
   * For example, if your image is called `background.png`, create a `background@2x.png` that is
   * double the size.
```

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+   */
+  background?: string;
+  /**
+   * Path to the icon to use for the app in the DMG window.
+   */
+  icon?: string;
+  /**
+   * How big to make the icon for the app in the DMG. [Default: `80`].
+   */
+   iconSize?: number;
```
Comment (erickzhao) [inline src/index.ts:34]: Is there linting set up for this repo? Seems like there's an extra space on this line.
```suggestion
  iconSize?: number;
```

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+} & ({
+  /**
+   * The directory to put the DMG into. [Default: `process.cwd()`].
+   */
+  out?: string;
+} | {
+  /**
+   * The full path to write the DMG to.
+   */
+  dmgPath: string;
```
Comment (erickzhao) [inline src/index.ts:73]: This should be an optional parameter I think
```suggestion
  dmgPath?: string;
```

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+   */
+  additionalDMGOptions?: Omit<appdmgType.Specification, 'title' | 'contents' | 'icon' | 'icon-size' | 'background' | 'format'>;
+} & ({
+  /**
+   * The directory to put the DMG into. [Default: `process.cwd()`].
+   */
+  out?: string;
+} | {
+  /**
+   * The full path to write the DMG to.
```
Comment (erickzhao) [inline src/index.ts:71]:
```suggestion
   * The full path to write the DMG to. This option cannot be specified at the same time as `out`.
```

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+  /**
+   * Additional options to pass through to [`appdmg`](https://npm.im/appdmg)
+   *
+   * You can use this to set additional features like `background-color` and
+   * `code-sign`.  See the docs of the `appdmg` module for all possible options.
+   */
+  additionalDMGOptions?: Omit<appdmgType.Specification, 'title' | 'contents' | 'icon' | 'icon-size' | 'background' | 'format'>;
+} & ({
+  /**
+   * The directory to put the DMG into. [Default: `process.cwd()`].
```
Comment (erickzhao) [inline src/index.ts:66]:
```suggestion
   * The directory to put the DMG into. This option cannot be specified at the same time as `dmgPath`.
   * 
   * Defaults to `process.cwd()`.
```

**Review by erickzhao — CHANGES_REQUESTED** [review]

**Comment by mahnunchik** [comment]
I'm looking forward to have it merged

**Review by erickzhao — APPROVED** [review]

File: src/electron-installer-dmg-bin.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  iconSize: args['icon-size'] ? parseInt(args['icon-size'], 10) : undefined,
+  background: args.background,
+  title: args.title,
+  format: args.format,
+  out: args.out,
+};
+
+if (args.help || args.h || !options.appPath || !options.name) {
+  console.error(usage);
+  process.exit(1);
```
Comment (felixrieseberg) [inline src/electron-installer-dmg-bin.ts:38]: I agree that we shouldn't change this in a TS refactor but this seems like an odd choice, `—help` usually ends in code 0

**Review by felixrieseberg — APPROVED** [review]

**Pull request merged**